File: backend/apps/pharma_engine/contraindications.py
"""Avaliação de contraindicações (bloqueios) antes do cálculo de dose.

Módulo puro Python (apenas Decimal), sem dependências do Django.

Diferente de ``limits.py`` (que emite avisos de dose fora da faixa), aqui o
resultado é um *bloqueio*: se uma regra casa com o paciente, a apresentação/via
escolhida não deve ser dosada. Os avisos seguem o mesmo formato estruturado dos
demais módulos, sempre com ``severity == "CRITICO"`` e ``type ==
"contraindicated"``.

Regras suportadas (ver docs/medication-schema.md):

- ``min_age_days``            -> idade mínima absoluta (em dias).
- ``min_weight_kg``           -> peso mínimo absoluto (em kg).
- ``form_min_age_days``       -> idade mínima para uma forma farmacêutica.
- ``route_min_age_days``      -> idade mínima para uma via.
- ``route_forbidden_age_range`` -> via proibida numa janela etária [min, max).
"""

from decimal import Decimal


def _block(message, *, drug="", rule="") -> dict:
    return {
        "type": "contraindicated",
        "severity": "CRITICO",
        "drug": drug,
        "rule": rule,
        "current_dose": "",
        "max_allowed": "",
        "unit": "",
        "message": message,
    }


def evaluate_contraindications(
    *,
    rules,
    age_days=None,
    weight=None,
    route=None,
    form=None,
    drug="",
) -> list[dict]:
    """Avalia as regras de contraindicação contra o paciente/seleção.

    Args:
        rules: lista de dicts (campo ``contraindications`` do catálogo).
        age_days: idade do paciente em dias (ou None se desconhecida).
        weight: peso em kg (ou None).
        route: via escolhida (ex: "IV"), quando aplicável.
        form: forma escolhida (ex: "supositorio"), quando aplicável.
        drug: nome do medicamento (para a mensagem).

    Returns:
        Lista de avisos de bloqueio (vazia se nenhuma regra casa).
    """
    if not rules:
        return []

    blocks: list[dict] = []

    for rule in rules:
        rtype = rule.get("rule")

        if rtype == "min_age_days":
            if age_days is not None and int(age_days) < int(rule["value"]):
                blocks.append(
                    _block(
                        f"Contraindicado: idade abaixo do mínimo "
                        f"({rule['value']} dias).",
                        drug=drug,
                        rule=rtype,
                    )
                )

        elif rtype == "min_weight_kg":
            if weight is not None and Decimal(str(weight)) < Decimal(
                str(rule["value"])
            ):
                blocks.append(
                    _block(
                        f"Contraindicado: peso abaixo do mínimo ({rule['value']} kg).",
                        drug=drug,
                        rule=rtype,
                    )
                )

        elif rtype == "form_min_age_days":
            if (
                form is not None
                and form == rule.get("form")
                and age_days is not None
                and int(age_days) < int(rule["value"])
            ):
                blocks.append(
                    _block(
                        f"Contraindicado: forma '{rule['form']}' abaixo da "
                        f"idade mínima ({rule['value']} dias).",
                        drug=drug,
                        rule=rtype,
                    )
                )

        elif rtype == "route_min_age_days":
            if (
                route is not None
                and route == rule.get("route")
                and age_days is not None
                and int(age_days) < int(rule["value"])
            ):
                blocks.append(
                    _block(
                        f"Contraindicado: via '{rule['route']}' abaixo da "
                        f"idade mínima ({rule['value']} dias).",
                        drug=drug,
                        rule=rtype,
                    )
                )

        elif rtype == "route_forbidden_age_range":
            if (
                route is not None
                and route == rule.get("route")
                and age_days is not None
                and int(rule["min_days"]) <= int(age_days) < int(rule["max_days"])
            ):
                blocks.append(
                    _block(
                        f"Contraindicado: via '{rule['route']}' nesta faixa "
                        f"etária ({rule['min_days']}-{rule['max_days']} dias).",
                        drug=drug,
                        rule=rtype,
                    )
                )

        else:
            raise ValueError(f"Regra de contraindicação desconhecida: {rtype}")

    return blocks

File: frontend/src/engines/calculator/contraindications.ts
import { Decimal } from '@/lib/decimal'

export interface ContraindicationWarning {
  type: string
  severity: string
  drug: string
  rule: string
  current_dose: string
  max_allowed: string
  unit: string
  message: string
}

function _block(message: string, drug = '', rule = ''): ContraindicationWarning {
  return {
    type: 'contraindicated',
    severity: 'CRITICO',
    drug,
    rule,
    current_dose: '',
    max_allowed: '',
    unit: '',
    message,
  }
}

export interface ContraindicationRule {
  rule: string
  value?: number
  form?: string
  route?: string
  min_days?: number
  max_days?: number
}

export function evaluateContraindications({
  rules,
  ageDays,
  weight,
  route,
  form,
  drug = '',
}: {
  rules: ContraindicationRule[] | null | undefined
  ageDays?: unknown
  weight?: unknown
  route?: string | null
  form?: string | null
  drug?: string
}): ContraindicationWarning[] {
  if (!rules || rules.length === 0) return []

  const blocks: ContraindicationWarning[] = []

  for (const rule of rules) {
    const rtype = rule.rule

    if (rtype === 'min_age_days') {
      if (rule.value == null) throw new Error(`Contraindication rule ${rtype} missing value`)
      if (ageDays != null && Math.floor(Number(ageDays)) < rule.value) {
        blocks.push(
          _block(
            `Contraindicado: idade abaixo do mínimo (${rule.value} dias).`,
            drug,
            rtype,
          ),
        )
      }
    } else if (rtype === 'min_weight_kg') {
      if (rule.value == null) throw new Error(`Contraindication rule ${rtype} missing value`)
      if (weight != null && new Decimal(String(weight)).lt(new Decimal(String(rule.value)))) {
        blocks.push(
          _block(
            `Contraindicado: peso abaixo do mínimo (${rule.value} kg).`,
            drug,
            rtype,
          ),
        )
      }
    } else if (rtype === 'form_min_age_days') {
      if (rule.value == null) throw new Error(`Contraindication rule ${rtype} missing value`)
      if (
        form != null &&
        form === rule.form &&
        ageDays != null &&
        Math.floor(Number(ageDays)) < rule.value
      ) {
        blocks.push(
          _block(
            `Contraindicado: forma '${rule.form}' abaixo da idade mínima (${rule.value} dias).`,
            drug,
            rtype,
          ),
        )
      }
    } else if (rtype === 'route_min_age_days') {
      if (rule.value == null) throw new Error(`Contraindication rule ${rtype} missing value`)
      if (
        route != null &&
        route === rule.route &&
        ageDays != null &&
        Math.floor(Number(ageDays)) < rule.value
      ) {
        blocks.push(
          _block(
            `Contraindicado: via '${rule.route}' abaixo da idade mínima (${rule.value} dias).`,
            drug,
            rtype,
          ),
        )
      }
    } else if (rtype === 'route_forbidden_age_range') {
      if (rule.min_days == null) throw new Error(`Contraindication rule ${rtype} missing min_days`)
      if (rule.max_days == null) throw new Error(`Contraindication rule ${rtype} missing max_days`)
      if (
        route != null &&
        route === rule.route &&
        ageDays != null &&
        Math.floor(Number(ageDays)) >= rule.min_days &&
        Math.floor(Number(ageDays)) < rule.max_days
      ) {
        blocks.push(
          _block(
            `Contraindicado: via '${rule.route}' nesta faixa etária (${rule.min_days}-${rule.max_days} dias).`,
            drug,
            rtype,
          ),
        )
      }
    } else {
      throw new Error(`Regra de contraindicação desconhecida: ${rtype}`)
    }
  }

  return blocks
}
